pjlib/src/pj/ssl_sock_ossl.c

```diff
@@ -455,6 +455,123 @@ static int openssl_init_count;
 /* OpenSSL application data index */
 static int sslsock_idx;
 
+#if defined(PJ_SSL_SOCK_OSSL_USE_THREAD_CB) && \
+    PJ_SSL_SOCK_OSSL_USE_THREAD_CB != 0 && OPENSSL_VERSION_NUMBER < 0x10100000L
+
+/* Thread lock pool.*/
+static pj_caching_pool 	 cp;
+static pj_pool_t 	*lock_pool;
+
+/* OpenSSL locking list. */
+static pj_lock_t **ossl_locks;
+
+/* OpenSSL number locks. */
+static unsigned ossl_num_locks;
+
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+static void ossl_set_thread_id(CRYPTO_THREADID *id)
+{
+    CRYPTO_THREADID_set_numeric(id,
+                     (unsigned long)pj_thread_get_os_handle(pj_thread_this()));
+}
+
+#else
+
+static unsigned long ossl_thread_id(void)
+{
+    return ((unsigned long)pj_thread_get_os_handle(pj_thread_this()));
+}
+#endif
+
+static void ossl_lock(int mode, int id, const char *file, int line)
+{
+    PJ_UNUSED_ARG(file);
+    PJ_UNUSED_ARG(line);
+
+    if (openssl_init_count == 0)
+        return;
+
+    if (mode & CRYPTO_LOCK) {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Lock File (%s) Line(%d)", file, line));
+            pj_lock_acquire(ossl_locks[id]);
+        }
+    } else {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Unlock File (%s) Line(%d)", file, line));
+            pj_lock_release(ossl_locks[id]);
+        }
+    }
+}
+
+static void release_thread_cb(void)
+{
+    unsigned i = 0;
+
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+    CRYPTO_THREADID_set_callback(NULL);
+#else
+    CRYPTO_set_id_callback(NULL);
+#endif
+    CRYPTO_set_locking_callback(NULL);
+
+    for (; i < ossl_num_locks; ++i) {
+        if (ossl_locks[i]) {
+            pj_lock_destroy(ossl_locks[i]);
+            ossl_locks[i] = NULL;
+        }
+    }
+    if (lock_pool) {
+        pj_pool_release(lock_pool);
+        lock_pool = NULL;
+    }
+    pj_caching_pool_destroy(&cp);
```

Comment:
just asking, is it okay to destroy a caching_pool that's never initialised?

Reply:
moved to above block.

Reply:
If moved to the block, it's the other way around, there is a possibility of the caching pool being initialised but never destroyed (it is an edge case though).
If I have to choose, I will still prefer the earlier version, i.e. destroying an uninitialised caching pool, provided that it's proven to be safe.

Reply:
the destructor will only be called if initializer is called (because atexit() is called by initializer), so there should be no destroy without init, right? but as atexit() may return error, destructor may not be called.

Reply:
right. shouldn't we call atexit() earlier, right after caching_pool_init()?

Reply:
the convention is to avoid occupying an atexit slot if possible. so atexit() should be called last after everything else goes well (just as is). OTOH, if any of the step fail, cleanup everything before returning error, print warning (please add), don't call atexit().